# "Annalena_RNAseq Pipeline": The pipeline to perform Differential Gene Expression Analysis and Functional Analysis for the RNASeq Data to Identify differences in the cytokine regulation and expression of immune cells of uninfected and Salmonella-infected animals at an age of 3 and 5 days. Additionally, the influence of the inflammatory signalling mediator MYD88 on the cytokine regulation should be examined.  . Wetlab data generated by Annalena Juergens.
## Author : Keshav Prasad Gubbi

# # Libraries
Package_List <-
  c(
    "dplyr",
    "DESeq2",
    "pheatmap",
    "PoiClaClu",
    "RColorBrewer",
    "vsn",
    "EnhancedVolcano",
    "gplots",
    "org.Mm.eg.db",
    "stringr",
    "genefilter",
    "tidyverse",
    "AnnotationDbi",
    "ComplexHeatmap",
    "DOSE",
    "clusterProfiler",
    "ggrepel",
    "GO.db",
    "GOstats",
    "gage",
    "gageData",
    "GOSemSim",
    "enrichplot",
    "ggnewscale",
    "glue",
    "ggupset",
    "FactoMineR",
    "factoextra",
    "here",
    "data.table",
    "ggrastr"
  )
not_installed <-
  Package_List[!(Package_List %in% installed.packages()[, "Package"])] # Extract not installed packages
if (length(not_installed)) {
  install.packages(not_installed)
} # Install the uninstalled packages
invisible(lapply(
  Package_List,
  suppressPackageStartupMessages(library),
  character.only = TRUE
))

# File Path Declarations
here::i_am(path = "Annalena_RNAseq.R")
here()

# Creating a comparison Variable: That Could be used later for all other comparison titles using a
# glue Variable. Define the Comparison and also create the folder for saving all plots and results to be
# saved as per the comparison
# ComparisonList <- c(
#   "Day3_UninfectedVSDay3_Infected", # SPF is the Numerator.
#   "Day5_UninfectedVSDay5_Infected", # d7 WT is the Numerator.
#   "Day3_UninfectedVSDay5_Uninfected", # d7 is the Numerator
#   "Day3_InfectedVSDay5_Infected", # SPF is the Numerator
# )
# for (i in ComparisonList) {
#   print(i)
#   Comparison <- i
#   print(Comparison)

Comparison <- "Day3_UninfectedVSDay3_Infected" # SPF is the Numerator.
# Comparison <- "Day5_UninfectedVSDay5_Infected" # d7 is the Numerator
# Comparison <- "Day3_UninfectedVSDay5_Uninfected"# SPF is the Numerator
# Comparison <- "Day3_InfectedVSDay5_Infected" # d7 WT is the Numerator

  # Determine the Comparison Condition: Comment one of them out based on the comparison you are trying to run.
  if (Comparison %in% c("Day3_UninfectedVSDay3_Infected", "Day5_UninfectedVSDay5_Infected")) {
    Comparison_Condition <- "Infection_Status"
    print(Comparison_Condition)
  } else {
    Comparison_Condition <- "Age_Days"
    print(Comparison_Condition)
  }
  # Comparison_Condition <- "MouseType"
  # Comparison_Condition <- "condition"

  # Folder Paths for Different Comparisons
  Comparison_path <- file.path(here(), glue("{Comparison}"))
  if (!dir.exists(here(Comparison_path))) {
    dir.create(here(Comparison_path))
  } else {
    print("Dir already exists!")
  }
  paste0(Comparison_path)

  # for Example: UninfectedVSInfected is the volcano plot notation: for Deseq2 results,
  # the infected is Numerator.and uninfected is Denominator. which will result in DESEQ contrast like (condition, Infected, Uninfected) and hence the infected part will come on the right side of the volcano plot.

# }


# Provide the countsmatrix and the samples
countsmatrix <- read.csv(file.path(here(), "Countmatrix_mapped_Anna_totalRNAseq.csv"))
# sorting and reordering names alphabetically
col_order <- c("symbols", "a1", "a2", "a3", "a4", "a5", "a6", "a7", "a8", "a9", "a10", "a11","a12", "a13", "a14","a15", "a16", "a17", "a18", "a19", "a20")
# Filter out the other 5 known gender genes as well from other RNAseq Projects.
countsmatrix <- countsmatrix[, col_order] %>% filter(symbols != "Xist",
                                                     symbols != "Jpx",
                                                     symbols != "Ftx",
                                                     symbols != "Tsx",
                                                     symbols != "Cnbp")
# Read in the list of Cytokine-specific genes supplied by Anna and Aline and keep rows only belonging to this list in the countmatrix.
cytokineList <- data.frame(read_excel("cytokine.xlsx"))
## Remove rows with only empty cells and rename the first column
cytokineList <- cytokineList[!apply(cytokineList == " ", MARGIN = 1, all), ] %>%
  drop_na() %>% rename(symbols = Symbol)

countsmatrix <- countsmatrix %>% inner_join(cytokineList,by = "symbols")# We can use the inner_join() function to return all rows in the first data frame that have a matching row in the second data frame
write.csv(countsmatrix, file.path(here(), "cytokine_filtered_countmatrix.csv"))# Save the respective cytokine filtered count matrix

# Removing the duplicated genes & then these genes put into rownames for countsmatrix and drop EnsemblId column
countsmatrix <- countsmatrix[!duplicated(countsmatrix$symbols), ] %>%
  remove_rownames() %>%
  column_to_rownames(var = "symbols") %>%
  select(-c(Name, Chr, Annotated.Term)) %>%
  as.matrix()

# Changing countsmatrix into Matrix of numeric values so that only numeric values are present in it as an input of DESEq Object.
class(countsmatrix) <- "numeric"

## Develop ColData by importing metadata
# Read the csv file and change the column name. the samples.csv is a list of sample names, ie, the names of bam files.
coldata <- as.data.frame(read.csv(file.path(here(), "Samples_anna.csv")))
colnames(coldata) <- c("Sample_Name", "Age_Days", "Infection_Status", "Annotated") # change name of one of the columns
rownames(coldata) <- coldata[, 4] # move the annotated sample names into row names
# The metadata can be found in a df called coldata!
## the elements from Sample_Name from coldata must match the the colnames of countsmatrix
colnames(countsmatrix) <- coldata$Annotated


### Reduce larger Matrix to smaller one - based on comparison
# *****************Now to the comparisons*************
# There are two groups of data here
# Day 3: Infected : c(5, 6, 7, 8) and Uninfected: c(9, 10, 13, 14)
# Day 5: Infected : c(1, 2, 3, 4) and Uninfected: c(11, 12, 15, 16)
### Reduce larger Matrix to smaller one - based on comparison
paste0(Comparison)
switch(Comparison,
       "Day3_UninfectedVSDay3_Infected" = {
         (coldata <- coldata[c(5, 6 , 7, 8, 9, 10, 13, 14), ])
       },
       "Day5_UninfectedVSDay5_Infected" = {
         (coldata <- coldata[c(1, 2, 3, 4, 11, 12, 15, 16), ])
       },
       "Day3_UninfectedVSDay5_Uninfected"  = {
         (coldata <- coldata[c(9, 10, 13, 14, 11, 12, 15, 16), ])
       },
       "Day3_InfectedVSDay5_Infected" = {
         (coldata <- coldata[c(1, 2, 3, 4, 5, 6, 7, 8), ])
       }
)
switch(Comparison,
       "Day3_UninfectedVSDay3_Infected" = {
         (countsmatrix <- countsmatrix[, c(5, 6 , 7, 8, 9, 10, 13, 14)])
       },
       "Day5_UninfectedVSDay5_Infected" = {
         (countsmatrix <- countsmatrix[, c(1, 2, 3, 4, 11, 12, 15, 16)])
       },
       "Day3_UninfectedVSDay5_Uninfected" = {
         (countsmatrix <- countsmatrix[, c(9, 10, 13, 14, 11, 12, 15, 16)])
       },
       "Day3_InfectedVSDay5_Infected" = {
         (countsmatrix <- countsmatrix[, c(1, 2, 3, 4, 5, 6, 7, 8)])
       }
)
paste0(coldata)
paste0(countsmatrix)
# **********************FUNCTIONS**********************************************
# Function to save generic plots
saveplot <- function(plot, plotname) {
  # Function to save the plots
  extension <- ".jpeg"
  ggsave(
    filename = file.path(Comparison_path, paste(
      plotname, glue("_{Comparison}_"), extension
    ), sep = ""),
    plot = plot,
    dpi = 300,
    width = 10,
    height = 10,
    units = "in"
  )
  # dev.off()
  while (!is.null(dev.list())) {
    dev.off()
  }
}
# **********************DESeq Analysis********************************
# Sanity Check for DDS
all(rownames(coldata) %in% colnames(countsmatrix))
ncol(countsmatrix) == nrow(coldata)
dim(countsmatrix)

# Create DEseq Object based on design that was chosen through the Comparison_Condition that was chosen at the begininning of the script run.
if (Comparison_Condition == "Infection_Status") {
  ## Creating the DESeq Data set Object
  dds <-
    DESeqDataSetFromMatrix(
      countData = countsmatrix,
      colData = coldata,
      design = ~Infection_Status
    )
} else {
  dds <-
    DESeqDataSetFromMatrix(
      countData = countsmatrix,
      colData = coldata,
      design = ~Age_Days
    )
}



# Further filtering of low count genes
# keep <- rowSums(counts(dds)) > 10
# dds <- dds[keep, ]
# nrow(dds)
## Applying VST transformation
vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
# vsd_coldata <- colData(vsd)
dds <- estimateSizeFactors(dds)


### Euclidean Distance between samples
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd$Sample_Name
colnames(sampleDistMatrix) <- vsd$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(
  EuclideanDistanceHeatmap <- pheatmap(
    sampleDistMatrix,
    clustering_distance_rows = sampleDists,
    clustering_distance_cols = sampleDists,
    main = glue("Euclidean Distance of Samples: {Comparison}"),
    col = colors
  )
)
### Poisson Distance between Samples
poisd <-
  PoissonDistance(t(counts(dds))) # raw counts or unnormalised data
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- dds$Sample_Name
colnames(samplePoisDistMatrix) <- dds$Sample_Name
colors <- colorRampPalette(rev(brewer.pal(9, "RdYlBu")))(255)
(
  poisson_dist_plot <- pheatmap(
    samplePoisDistMatrix,
    clustering_distance_rows = poisd$dd,
    clustering_distance_cols = poisd$dd,
    main = glue("Poisson Distance of Samples: {Comparison}"),
    col = colors
  )
)
#  **************************PCA Plot**********************************
# Functions for Plot aethetics and saving PCA Plots
color_values <-
  c(
    "black",
           "black",
           "red",
           "red",
           "red",
           "red",
           "red",
           "red",
           "red",
           "red",
           "black",
           "black"
  )
# The basic set of common aesthetic settings for PCA plots,
theme.my.own <- list(
  theme_bw(),
  geom_point(size = 3),
  coord_fixed(),
  scale_y_continuous(
    breaks = seq(-100, 100, 10),
    sec.axis = sec_axis(~ . * 1, labels = NULL, breaks = NULL)
  ),
  scale_x_continuous(
    breaks = seq(-50, 50, 10),
    sec.axis = sec_axis(~ . * 1, labels = NULL, breaks = NULL)
  ),
  theme_classic(),
  geom_hline(
    yintercept = 0,
    color = "gray",
    linewidth = 1
  ),
  geom_vline(
    xintercept = 0,
    color = "gray",
    linewidth = 1
  ),
  theme(
    text = element_text(size = 15),
    axis.text = element_text(size = 15),
    legend.position = "bottom",
    aspect.ratio = 1
  ),
  # geom_text(size = 4, hjust = 0, vjust = 0),
  geom_text_repel(size = 4, min.segment.length = 0.1)
)
# PCA Plot Calculation
# Calculating all PCA Values
plotPCA_local <-
  function(object,
           intgroup = "condition",
           ntop = 500,
           returnData = TRUE,
           nPC = 4) {
    # calculate the variance for each gene
    rv <- rowVars(assay(object))
    ntop <- 500
    # select the ntop genes by variance
    select <-
      order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
    # perform a PCA on the data in assay(x) for the selected genes
    pca <- prcomp(t(assay(object)[select, ]))
    # summary(pca)
    # the contribution to the total variance for each component
    percentVar <- pca$sdev^2 / sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
      stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <-
      as.data.frame(colData(object)[, intgroup, drop = FALSE])
    # add the intgroup factors together to create a new grouping factor
    group <- if (length(intgroup) > 1) {
      factor(apply(intgroup.df, 1, paste, collapse = ":"))
    } else {
      colData(object)[[intgroup]]
    }
    # assembly the data for the plot
    d <-
      cbind(
        pca$x[, seq_len(min(nPC, ncol(pca$x))), drop = FALSE],
        data.frame(group = group, intgroup.df, name = colnames(object))
      )
    if (returnData) {
      attr(d, "percentVar") <- percentVar[1:nPC]
      # l <- list(pca,d)
      # return(l)
      return(d)
    }
  }
## PCA Plot with VST Data
### Function for calculating percentvar for different variables
percentvar_calculation <- function(pcaData_variable) {
  percentvar_variable <-
    round(100 * attr(pcaData_variable, "percentVar"), digits = 3)
  return(percentvar_variable)
}

switch(Comparison_Condition,
       "Infection_Status" = {
         pcaData <- plotPCA_local(vsd,
                                  intgroup = c("Infection_Status", "Annotated"),
                                  returnData = T
         )
       }, #
       "Age_Days" = {
         pcaData <- plotPCA_local(vsd,
                                  intgroup = c("Age_Days", "Annotated"),
                                  returnData = T
         )
       },
)
print(pcaData)
percentVar <- percentvar_calculation(pcaData)
print(percentVar)

# PC Plot: PC1 vs PC2
(
  PCAplot_vst <-
    ggplot(
      pcaData,
      aes(
        x = PC1,
        y = PC2,
        color = Annotated,
        label = Annotated
      )
    ) +
    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    ggtitle(glue("PCA: {Comparison}")) +
    # scale_colour_manual(values = color_values) +
    theme.my.own
)
saveplot(PCAplot_vst, plotname = "PCA_PC1vsPC2")

# PCA Plot : PC3 vs PC4
(
  PCAplot_pc34 <- ggplot(
    pcaData,
    aes(
      x = PC3,
      y = PC4,
      color = Annotated,
      label = Annotated
    )
  ) +
    xlab(paste0("PC3: ", percentVar[3], "% variance")) +
    ylab(paste0("PC4: ", percentVar[4], "% variance")) +
    ggtitle(glue("PCA: {Comparison}")) +
    # scale_colour_manual(values = color_values) +
    theme.my.own
)
saveplot(PCAplot_pc34, plotname = "PCA_PC3vsPC4")

# ************************FactoExtra************************
# calculate the variance for top 500 gene

rv <- rowVars(assay(vsd))
ntop <- 500
# select the ntop genes by variance
select <-
  order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
res.pca <-
  PCA(t(assay(vsd)[select, ]), graph = FALSE, scale.unit = FALSE)
summary.PCA(res.pca)

# Visualize eigenvalues/variances
fviz_screeplot(res.pca, addlabels = TRUE)
eig.val <- get_eigenvalue(res.pca)
var <- get_pca_var(res.pca)

## Genes + PCA Biplots
heat.colors <- brewer.pal(6, "RdYlBu")
## Genes + PCA Biplots
(Genes_Biplot <- fviz_pca_biplot(res.pca, repel = TRUE))
saveplot(Genes_Biplot, plotname = "Genes_Biplot")

(
  Genes_contributions_Biplot <-
    fviz_pca_var(
      res.pca,
      col.var = "contrib",
      repel = TRUE,
      gradient.cols = c(
        "gray",
              "blue",
              "pink",
              "magenta",
              "yellow",
              "orange",
              "brown",
              "green",
              "red",
              "black"
      )
    )
)
saveplot(Genes_contributions_Biplot, plotname = "Genes_contributions_Biplot")
# Contributions of variables to PC2
(top25_genes_dim2 <-
    fviz_contrib(
      res.pca,
      choice = "var",
      axes = 2,
      top = 25
    ))
saveplot(top25_genes_dim2, plotname = "top25_genes_dim2")
# # Contributions of variables to PC1
(top25_genes_dim1 <-
    fviz_contrib(
      res.pca,
      choice = "var",
      axes = 1,
      top = 25
    ))
saveplot(top25_genes_dim1, plotname = "top25_genes_dim1")